File: java/it/eja/surf/Setting.java
package it.eja.surf;

import android.util.Base64;

import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONException;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.net.URL;
import java.nio.charset.StandardCharsets;

public class Setting {
    public static String host = "eja.surf";
    public static String home = "";
    public static String path = "";
    public static String doh = "";
    public static int version = 0;
    public static boolean reset = false;
    public static JSONObject eja = new JSONObject();
    public static JSONArray book = new JSONArray();
    public static JSONArray allow = new JSONArray();
    public static JSONArray block = new JSONArray();

    public static String load() throws JSONException {
        String data = fileRead("eja.json");
        if (data.length() > 0) {
            eja = new JSONObject(data);
        } else {
            eja.put("host", host);
            Setting.save();
        }
        eja.put("version", version);
        if (eja.has("host")) {
            host = eja.getString("host");
        }
        if (eja.has("reset")) {
            reset = eja.getBoolean("reset");
        }
        if (eja.has("doh")) {
            doh = eja.getString("doh");
        }
        if (Setting.eja.has("proxy") && Setting.eja.getBoolean("proxy")) {
            if (Setting.eja.has("socksHost") && Setting.eja.has("socksPort")) {
                System.setProperty("socksProxyHost", Setting.eja.getString("socksHost"));
                System.setProperty("socksProxyPort", Setting.eja.getString("socksPort"));
            }
        }
        data = fileRead("eja.book");
        if (data.length() > 0) {
            book = new JSONArray(data);
        }
        if (eja.has("allow") && !eja.getString("allow").isEmpty()) {
            data = fileRead("eja.allow");
            if (data.length() > 0) {
                allow = new JSONArray(data);
            }
        }
        if (eja.has("block") && !eja.getString("block").isEmpty()) {
            data = fileRead("eja.block");
            if (data.length() > 0) {
                block = new JSONArray(data);
            }
        }
        home = String.format("https://%s/", host);
        return eja.toString();
    }

    public static void save() {
        fileWrite("eja.json", eja.toString());
    }

    public static void bookAdd(String value) {
        if (value.toLowerCase().startsWith("http")) {
            book.put(value);
            fileWrite("eja.book", book.toString());
        }
    }

    public static void bookRemove(Integer index) {
        book.remove(index);
        fileWrite("eja.book", book.toString());
    }

    public static String fileRead(String fileName) {
        StringBuilder contentBuilder = new StringBuilder();
        String filePath = String.format("%s%s%s", path, File.separator, fileName);
        File file = new File(filePath);
        if (file.exists()) {
            try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
                String sCurrentLine;
                while ((sCurrentLine = br.readLine()) != null) {
                    contentBuilder.append(sCurrentLine).append("\n");
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return contentBuilder.toString();
    }

    public static void fileWrite(String fileName, String value) {
        String filePath = String.format("%s%s%s", path, File.separator, fileName);
        Writer writer;
        try {
            writer = new BufferedWriter(new OutputStreamWriter(
                    new FileOutputStream(filePath)));
            writer.write(value);
            writer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static boolean checkList(JSONArray list, String value) {
        for (int i = 0; i < list.length(); i++) {
            try {
                if (value.equals(list.getString(i))) {
                    return true;
                } else if (list.getString(i).startsWith("*") && value.endsWith(list.getString(i).substring(1))) {
                    return true;
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return false;
    }

    public static String dohToIp(String domain) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream query = new DataOutputStream(baos);
        String ip = "0.0.0.0";
        try {
            int id = (int) (Math.random() * 0xffff);
            query.writeShort(id);
            query.writeShort(0x0100);
            query.writeShort(0x0001);
            query.writeShort(0x0000);
            query.writeShort(0x0000);
            query.writeShort(0x0000);
            String[] levels = domain.split("\\.");
            for (String level : levels) {
                byte[] bytes = level.getBytes(StandardCharsets.UTF_8);
                query.writeByte(bytes.length);
                query.write(bytes);
            }
            query.writeByte(0x00);
            query.writeShort(0x0001);
            query.writeShort(0x0001);
            byte[] br = new byte[1024];
            byte[] bw = baos.toByteArray();
            InputStream in = new URL(String.format("%s?dns=%s", Setting.doh, Base64.encodeToString(bw, Base64.URL_SAFE | Base64.NO_PADDING | Base64.NO_WRAP))).openStream();
            int l = in.read(br);
            int x = bw.length;
            if (br[0] == bw[0] && br[1] == bw[1]) {
                while (x < l && br[x] == (byte) 0xc0) {
                    if (br[x + 3] == 0x01 && br[x + 11] == 4) {
                        ip = String.format("%d.%d.%d.%d", br[x + 12] & 0xff, br[x + 13] & 0xff, br[x + 14] & 0xff, br[x + 15] & 0xff);
                        break;
                    } else {
                        x = x + 11 + 1 + br[x + 11];
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return ip;
    }
}

File: app/src/main/java/it/eja/surf/Setting.kt
// Copyright (C) by Ubaldo Porcheddu <ubaldo@eja.it>

package it.eja.surf

import android.util.Base64
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.io.File
import java.io.IOException
import java.net.URL
import java.nio.charset.StandardCharsets

object Setting {
    var home = "http://surf.eja.it"
    var doh = "http://surf.eja.it"
    var path = ""
    var version = 0
    var reset = true
    var eja = JSONObject()
    var book = JSONArray()

    fun load() {
        val data = readInternal("eja.json")
        if (data.isNotEmpty()) {
            eja = JSONObject(data)
        } else {
            eja.put("home", home)
            save()
        }

        home = if (home.startsWith("http")) home else "http://$home"

        reset = eja.optBoolean("reset", reset)
        doh = eja.optString("doh", doh)

        if (eja.optBoolean("proxy", false)) {
            val sHost = eja.optString("socksHost")
            val sPort = eja.optString("socksPort")
            if (sHost.isNotEmpty()) System.setProperty("socksProxyHost", sHost)
            if (sPort.isNotEmpty()) System.setProperty("socksProxyPort", sPort)
        }

        book = eja.optJSONArray("bookmarks") ?: JSONArray()
    }

    fun save() {
        eja.put("bookmarks", book)
        writeInternal("eja.json", eja.toString())
    }

    fun bookAdd(value: String) {
        for (i in 0 until book.length()) {
            if (book.getString(i) == value) return
        }
        book.put(value)
        save()
    }

    fun bookEdit(index: Int, newValue: String) {
        if (index >= 0 && index < book.length()) {
            book.put(index, newValue)
            save()
        }
    }

    fun bookRemove(index: Int) {
        if (index >= 0 && index < book.length()) {
            book.remove(index)
            save()
        }
    }

    fun bookSwap(i: Int, j: Int) {
        if (i in 0 until book.length() && j in 0 until book.length()) {
            val temp = book.getString(i)
            book.put(i, book.getString(j))
            book.put(j, temp)
            save()
        }
    }

    private fun readInternal(fileName: String): String {
        val filePath = "$path${File.separator}$fileName"
        val file = File(filePath)
        if (!file.exists()) return ""
        return try {
            file.readText()
        } catch (e: IOException) {
            ""
        }
    }

    private fun writeInternal(fileName: String, value: String) {
        val filePath = "$path${File.separator}$fileName"
        try {
            File(filePath).writeText(value)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun dohToIp(domain: String): String {
        val baos = ByteArrayOutputStream()
        val query = DataOutputStream(baos)
        var ip = "0.0.0.0"
        try {
            val id = (Math.random() * 0xffff).toInt()
            query.apply {
                writeShort(id)
                writeShort(0x0100)
                writeShort(0x0001)
                writeShort(0x0000)
                writeShort(0x0000)
                writeShort(0x0000)
                domain.split('.').forEach { level ->
                    writeByte(level.length)
                    write(level.toByteArray(StandardCharsets.UTF_8))
                }
                writeByte(0x00)
                writeShort(0x0001)
                writeShort(0x0001)
            }

            val encodedQuery = Base64.encodeToString(baos.toByteArray(), Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
            val url = URL("$doh?dns=$encodedQuery")

            url.openStream().use { input ->
                val br = ByteArray(1024)
                val bw = baos.toByteArray()
                val bytesRead = input.read(br)

                var x = bw.size
                if (br[0] == bw[0] && br[1] == bw[1]) {
                    while (x < bytesRead && br[x] == 0xc0.toByte()) {
                        if (br[x + 3].toInt() == 0x01 && br[x + 11].toInt() == 4) {
                            ip = "${br[x + 12].toInt() and 0xff}.${br[x + 13].toInt() and 0xff}.${br[x + 14].toInt() and 0xff}.${br[x + 15].toInt() and 0xff}"
                            break
                        } else {
                            x += 11 + 1 + br[x + 11]
                        }
                    }
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return ip
    }
}